src/dataframerow/utils.jl

```diff
@@ -70,36 +70,30 @@ function hashrows_col!(h::Vector{UInt},
 end
 
 # Calculate the vector of `df` rows hash values.
-function hashrows(df::AbstractDataFrame, skipmissing::Bool)
-    rhashes = zeros(UInt, nrow(df))
-    missings = fill(false, skipmissing ? nrow(df) : 0)
-    for (i, col) in enumerate(columns(df))
+function hashrows(cols::Tuple{Vararg{AbstractVector}}, skipmissing::Bool)
+    len = length(cols[1])
+    rhashes = zeros(UInt, len)
+    missings = fill(false, skipmissing ? len : 0)
+    for (i, col) in enumerate(cols)
         hashrows_col!(rhashes, missings, col, i == 1)
     end
     return (rhashes, missings)
 end
 
 # Helper function for RowGroupDict.
 # Returns a tuple:
-# 1) the number of row groups in a data table
-# 2) vector of row hashes
+# 1) the number of row groups in a data frame
+# 2) vector of row hashes (may be empty if hash=Val(false))
 # 3) slot array for a hash map, non-zero values are
 #    the indices of the first row in a group
 # Optional group vector is set to the group indices of each row
-function row_group_slots(df::AbstractDataFrame,
-                         groups::Union{Vector{Int}, Nothing} = nothing,
-                         skipmissing::Bool = false)
-    rhashes, missings = hashrows(df, skipmissing)
-    row_group_slots(ntuple(i -> df[i], ncol(df)), rhashes, missings, groups, skipmissing)
-end
-
 function row_group_slots(cols::Tuple{Vararg{AbstractVector}},
-                         rhashes::AbstractVector{UInt},
-                         missings::AbstractVector{Bool},
+                         hash::Val = Val(true),
                          groups::Union{Vector{Int}, Nothing} = nothing,
-                         skipmissing::Bool = false)
+                         skipmissing::Bool = false)::Tuple{Int, Vector{UInt}, Vector{Int}, Bool, Bool}
```

Comment:
why is this needed?

Reply:
Yeah, that's annoying, it's because we call it from `group_rows`, which takes a `DataFrame`, and therefore doesn't know the types of the columns it passes to `row_group_slots`. Since all `row_group_slots` return the same argument types, the compiler should be able to figure the return type anyway, but that doesn't seem to work without the assertions.